ensime_shared/protocol.py

```diff
@@ -197,54 +193,52 @@ def handle_string_response(self, call_id, payload):
           2. `DebugToStringReq`
           3. `FormatOneSourceReq`
         """
-        self.log(str(payload))
+        self.log.debug('handle_string_response: in %s', payload)
         self.handle_doc_uri(call_id, payload)
 
     def handle_doc_uri(self, call_id, payload):
         """Handler for responses of Doc URIs."""
-        if not self.en_format_source_id:
-            self.log("handle_string_response: received doc path")
+        if not self.en_format_source_id:  # TODO: what does format source have to do with this?
+            self.log.debug('handle_doc_uri: received doc path')
             port = self.ensime.http_port()
 
             url = payload["text"]
 
             if not url.startswith("http"):
                 url = gconfig["localhost"].format(port, payload["text"])
 
-            browse_enabled = self.call_options[call_id].get("browse")
-
-            if browse_enabled:
-                log_msg = "handle_string_response: browsing doc path {}"
-                self.log(log_msg.format(url))
+            if self.call_options[call_id].get('browse'):
+                self.log.debug('handle_doc_uri: browsing doc path %s', url)
                 try:
                     if webbrowser.open(url):
-                        self.log("opened {}".format(url))
-                except webbrowser.Error as e:
-                    log_msg = "handle_string_response: webbrowser error: {}"
-                    self.log(log_msg.format(e))
-                    self.raw_message(feedback["manual_doc"].format(url))
+                        self.log.info('opened %s', url)
+                except webbrowser.Error:
+                    self.log.exception('handle_doc_uri: webbrowser error')
+                    self.editor.raw_message(feedback["manual_doc"].format(url))
 
             del self.call_options[call_id]
         else:
+            # TODO: Get rid of this self.vim usage, but I don't understand
+            # when/if we ever get here...
             self.vim.current.buffer[:] = \
                 [line.encode('utf-8') for line in payload["text"].split("\n")]
             self.en_format_source_id = None
```

Comment:
What I understand after a quick read of the source code:

This method `handle_doc_uri`, indirectly through `handle_string_response` is used to handle all responses of type `StringResponse` (very tricky, good candidate to be refactored #330)

Currently there is two very different kind of message handle by this method:

- __DocBrowse__: handled in the `True` branch of the `if` where all the _open-with-browser-magic_ is done
- __FormatSource__: handled in the `else` branch under question, you get here with the __:EnFormatSource__ command which sends the message that is responded with a `StringResponse` containing the formatted source code. This command also sets `en_format_source_id` to a __True-like__ value in [ensime.py:397](https://github.com/ches/ensime-vim/blob/8db528a32e7466b7a9df2326d75764d4c8dc4085/ensime_shared/ensime.py#L397). The text of the `current.buffer` is replaced with the one from the response using the `vim` instance. So you could move this to the `Editor Object` as something like `editor.replaceText(payload['text'])` which I think is really dangerours since the request can take a few seconds to be responded and the current buffer could potentially be a different one. IMO this feature should be safely implemented or removed.